File: apps/multimodal-screener/iOS/Aiberry/Face/FaceEmotionModel.swift
import Foundation
import CoreVideo
import ZeticMLange

/// Facial-expression recognition on-device via a ViT FER model on Melange.
/// Input: `1 × 3 × 224 × 224` raw RGB; output: 7 logits in `AppConfig.emotionLabels`
/// order (the export wrapper reorders to the canonical order — see prepare script).
///
/// Fed throttled frames from `CameraController`. Detection (Apple Vision) + crop +
/// inference all run on a private queue with a single-in-flight gate, and per-frame
/// softmax probabilities are accumulated into a session-wide running mean — robust
/// to the odd blurry/occluded frame (cf. the voice model's multi-window averaging).
final class FaceEmotionModel: ObservableObject {
    @Published var status: ModelStatus = .idle
    @Published var liveScores: [EmotionScore] = []   // running-mean distribution, ranked
    @Published var framesWithFace = 0
    @Published var latencyMs: Double?

    private var model: ZeticMLangeModel?
    private let queue = DispatchQueue(label: "aiberry.face")
    private let detector = FaceDetector()
    // Single-in-flight gate. Checked synchronously in `ingest` so a new frame is
    // dropped *before* it's captured into a queued closure — bounds work to one
    // inference at a time and prevents camera buffers piling up in memory.
    private let gate = NSLock()
    private var busy = false

    private let labels = AppConfig.emotionLabels        // canonical order (output/accumulator)
    private let modelLabels = AppConfig.faceModelLabels // hosted model's native logit order
    /// modelLabels[i] → canonical index, so we can read native logits and accumulate
    /// in canonical order (keeps fusion/styling model-agnostic).
    private let toCanonical: [Int]
    private var summed: [Float]                          // running sum of per-frame softmax
    private var frames = 0

    var liveTop: EmotionScore? { liveScores.max(by: { $0.probability < $1.probability }) }

    init() {
        summed = [Float](repeating: 0, count: AppConfig.emotionLabels.count)
        toCanonical = AppConfig.faceModelLabels.map { native in
            let canonical = AppConfig.faceLabelToCanonical[native] ?? native
            return AppConfig.emotionLabels.firstIndex(of: canonical) ?? 0
        }
    }

    /// Download + compile the model ahead of time (called at launch).
    func preload() {
        guard model == nil, !status.isBusy else { return }
        status = .downloading(0)
        queue.async { [weak self] in
            guard let self else { return }
            do { _ = try self.ensureModel(); DispatchQueue.main.async { self.status = .idle } }
            catch { DispatchQueue.main.async { self.status = .failed(String(describing: error)) } }
        }
    }

    /// Begin a fresh accumulation for a new check-in.
    func reset() {
        queue.async { [weak self] in
            guard let self else { return }
            self.summed = [Float](repeating: 0, count: self.labels.count)
            self.frames = 0
            DispatchQueue.main.async {
                self.liveScores = []
                self.framesWithFace = 0
                self.latencyMs = nil
            }
        }
    }

    /// Ingest one camera frame. Dropped if a previous frame is still being processed
    /// or if the model isn't ready. Frames with no detected face are skipped.
    func ingest(_ pixelBuffer: CVPixelBuffer) {
        // Drop now (without capturing the buffer) if an inference is already running.
        gate.lock()
        if busy { gate.unlock(); return }
        busy = true
        gate.unlock()

        queue.async { [weak self] in
            guard let self else { return }
            defer { self.gate.lock(); self.busy = false; self.gate.unlock() }
            guard let model = self.model else { return }

            guard let face = self.detector.detect(pixelBuffer) else { return }
            let values = FacePixelTensor.bgrMeanSubtracted(face.rgba, size: face.size)
            let size = face.size
            let input = MelangeKit.floatTensor(values, shape: [1, 3, size, size])
            do {
                let (outputs, ms) = try measureMs { try model.run(inputs: [input]) }
                guard let logits = outputs.first.map({ MelangeKit.floats(from: $0) }),
                      logits.count >= self.modelLabels.count else { return }
                // Softmax over the model's native order, then accumulate in canonical order.
                let probs = MelangeKit.softmax(Array(logits.prefix(self.modelLabels.count)))
                // Weight each frame by detection confidence so a marginal detection
                // counts less toward the running mean.
                let w = max(0.2, face.confidence)
                for i in 0..<self.modelLabels.count { self.summed[self.toCanonical[i]] += probs[i] * w }
                self.frames += 1
                let mean = self.summed.map { $0 / Float(self.frames) }
                let ranked = zip(self.labels, mean)
                    .map { EmotionScore(label: $0.0, probability: $0.1) }
                    .sorted { $0.probability > $1.probability }
                DispatchQueue.main.async {
                    self.liveScores = ranked
                    self.framesWithFace = self.frames
                    self.latencyMs = ms
                }
            } catch {
                DispatchQueue.main.async { self.status = .failed(String(describing: error)) }
            }
        }
    }

    /// The session-wide face distribution (probabilities by `AppConfig.emotionLabels`),
    /// plus how many good frames it was built from. Nil if no face was ever seen.
    func finalize(_ completion: @escaping (_ distribution: [Float], _ framesWithFace: Int) -> Void) {
        queue.async { [weak self] in
            guard let self else { return }
            let f = self.frames
            guard f > 0 else { DispatchQueue.main.async { completion([], 0) }; return }
            // Renormalize the (confidence-weighted) running sum to a probability vector.
            let total = self.summed.reduce(0, +)
            let dist = total > 0 ? self.summed.map { $0 / total } : self.summed
            DispatchQueue.main.async { completion(dist, f) }
        }
    }

    private func ensureModel() throws -> ZeticMLangeModel {
        if let model { return model }
        let loaded = try MelangeKit.load(AppConfig.Model.face, version: AppConfig.Model.faceVersion) { progress in
            DispatchQueue.main.async { self.status = .downloading(progress) }
        }
        self.model = loaded
        return loaded
    }
}

File: apps/multimodal-screener/Android/app/src/main/java/ai/zetic/aiberry/face/FaceEmotionModel.kt
package ai.zetic.aiberry.face

import android.content.Context
import android.graphics.Bitmap
import android.graphics.Matrix
import android.os.Handler
import android.os.Looper
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableIntStateOf
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.setValue
import ai.zetic.aiberry.core.AppConfig
import ai.zetic.aiberry.core.MelangeKit
import ai.zetic.aiberry.core.MelangeRuntime
import ai.zetic.aiberry.core.ModelStatus
import ai.zetic.aiberry.core.measureMs
import ai.zetic.aiberry.emotion.EmotionScore
import com.zeticai.mlange.core.model.ZeticMLangeModel
import java.util.concurrent.Callable
import java.util.concurrent.Executors
import java.util.concurrent.atomic.AtomicBoolean
import kotlin.math.max

/**
 * Live facial-expression recognition via `ElenaRyumina/FaceEmotionRecognition` (v1).
 *
 * Per frame (throttled to ~3 Hz by [ai.zetic.aiberry.face.CameraController]): ML Kit detects +
 * crops the face -> [FacePixelTensor] builds the BGR/mean-subtracted `[1,3,224,224]` tensor ->
 * the model emits 7 logits in its native order -> softmax -> remapped to canonical order ->
 * folded into a confidence-weighted running mean. 1:1 with iOS-Aiberry's `FaceEmotionModel`.
 *
 * Detection/crop run on a private single thread; the Melange `run` is funnelled onto the
 * shared [MelangeRuntime.executor] (the SDK forbids concurrent model init/run).
 */
class FaceEmotionModel(private val context: Context) {
    var status by mutableStateOf<ModelStatus>(ModelStatus.Idle)
        private set
    var liveScores by mutableStateOf<List<EmotionScore>>(emptyList())
        private set
    var framesWithFace by mutableIntStateOf(0)
        private set
    var latencyMs by mutableStateOf<Double?>(null)
        private set

    private var model: ZeticMLangeModel? = null
    private val melange = MelangeRuntime.executor          // shared; model load + run only
    private val faceWork = Executors.newSingleThreadExecutor { r ->
        Thread(r, "aiberry-face").apply { isDaemon = true }
    }
    private val main = Handler(Looper.getMainLooper())
    private val detector = FaceDetector()
    private val busy = AtomicBoolean(false)

    private val labels = AppConfig.emotionLabels
    // native logit index -> canonical [emotionLabels] index
    private val nativeToCanonical = IntArray(7) { i ->
        labels.indexOf(AppConfig.Face.labelToCanonical[AppConfig.Face.nativeLabels[i]])
    }

    // Running sum (canonical order) of softmax*weight, and the good-frame count.
    private val summed = FloatArray(7)
    private var frames = 0

    val liveTop: EmotionScore? get() = liveScores.maxByOrNull { it.probability }

    fun preload() {
        if (model != null || status.isBusy) return
        status = ModelStatus.Downloading(0f)
        melange.execute {
            try {
                ensureModel()
                main.post { if (status !is ModelStatus.Failed) status = ModelStatus.Idle }
            } catch (e: Throwable) {
                main.post { status = ModelStatus.Failed(MelangeKit.friendly(e)) }
            }
        }
    }

    /** Clear the running accumulator at the start of a new check-in. */
    fun reset() {
        faceWork.execute {
            summed.fill(0f)
            frames = 0
            main.post {
                liveScores = emptyList()
                framesWithFace = 0
            }
        }
    }

    /**
     * Feed one camera frame. [bitmap] is the latest analysis frame (RGBA), [rotationDegrees]
     * its orientation. Dropped immediately if a previous frame is still being processed
     * (single-in-flight gate), so the live stream never piles up.
     */
    fun ingest(bitmap: Bitmap, rotationDegrees: Int) {
        if (model == null) return
        if (!busy.compareAndSet(false, true)) return
        faceWork.execute {
            try {
                val m = model ?: return@execute
                val upright = rotateUpright(bitmap, rotationDegrees)
                val face = detector.detect(upright) ?: return@execute
                val values = FacePixelTensor.bgrMeanSubtracted(face.bitmap)
                val input = MelangeKit.floatTensor(values, intArrayOf(1, 3, 224, 224))
                val (outputs, ms) = measureMs {
                    melange.submit(Callable { m.run(arrayOf(input)) }).get()
                }
                val logits = outputs.firstOrNull()?.let { MelangeKit.floats(it) } ?: return@execute
                if (logits.size < 7) return@execute
                val probs = MelangeKit.softmax(logits.copyOf(7))

                val w = max(0.2f, face.confidence)
                for (i in 0 until 7) summed[nativeToCanonical[i]] += probs[i] * w
                frames += 1

                val mean = FloatArray(7) { summed[it] / frames }
                val ranked = labels.indices
                    .map { EmotionScore(labels[it], mean[it]) }
                    .sortedByDescending { it.probability }
                val f = frames
                main.post {
                    liveScores = ranked
                    framesWithFace = f
                    latencyMs = ms
                }
            } catch (_: Throwable) {
                // Drop this frame; the next one will try again.
            } finally {
                busy.set(false)
            }
        }
    }

    /**
     * End-of-session distribution: the running sum renormalized to a probability vector in
     * canonical order, plus the good-frame count. Delivered on the main thread.
     */
    fun finalize(onResult: (distribution: FloatArray, framesWithFace: Int) -> Unit) {
        faceWork.execute {
            val f = frames
            if (f <= 0) {
                main.post { onResult(FloatArray(0), 0) }
                return@execute
            }
            val total = summed.sum()
            val dist = if (total > 0f) FloatArray(7) { summed[it] / total } else summed.copyOf()
            main.post { onResult(dist, f) }
        }
    }

    @Synchronized
    private fun ensureModel(): ZeticMLangeModel {
        model?.let { return it }
        val loaded = MelangeKit.load(
            context,
            AppConfig.Model.FACE,
            AppConfig.Model.FACE_VERSION,
        ) { progress -> main.post { status = ModelStatus.Downloading(progress) } }
        model = loaded
        return loaded
    }

    private fun rotateUpright(bitmap: Bitmap, rotationDegrees: Int): Bitmap {
        if (rotationDegrees % 360 == 0) return bitmap
        val m = Matrix().apply { postRotate(rotationDegrees.toFloat()) }
        return Bitmap.createBitmap(bitmap, 0, 0, bitmap.width, bitmap.height, m, true)
    }
}
